### PR TITLE
[NPU] Sync Ascend with main and consolidate NPU adaptations

### PR DESCRIPTION
**Pull request by Meihan-chen**
## Summary

Bring Ascend up to date with main through `ce92eff1`, including the new native
model-loading, weight-transfer and accelerator interfaces, while incorporating
Ascend's training-stack updates and GLM-4.7-Flash support through `083aea63`.

The integration follows main's shared training and rollout flow. Necessary NPU
differences are kept in small platform/provider interfaces and
`docker/npu_patch`, rather than maintaining a separate copy of main's
orchestration. This keeps subsequent main merges easier to maintain.

## Main changes and corresponding NPU adaptations

| Area | Changes from main | NPU adaptation in this PR |
| --- | --- | --- |
| Weight synchronization | Native stateful trainer/worker weight-transfer interfaces replace the legacy worker-extension path; colocated transfer uses packed IPC payloads. | Reuse the shared update sequence, bucket packing and per-engine-slot coordination. Select `hccl` / `npu_ipc` and their initialization types through the NPU provider; implement the matching native worker contracts in vendor patches and remove obsolete Vime worker extensions. |
| MoE weight transfer | Rank-local expert routing and payloads avoid treating every rank as a full-model weight source. | Support the corresponding HCCL/IPC payloads, list workers and empty payload boundaries; preserve MoE loader metadata and routing/control buffers during reload. |
| Model loading and Bridge | Native HF-to-Megatron loading and Megatron-to-HF export replace the Bridge-based model integration and legacy conversion-mode arguments. | Follow the native loader/exporter structure for Qwen3-VL and GLM-4.7-Flash, including multimodal configuration, weight-prefix mapping and MLA/MoE/MTP mapping coverage. Do not reintroduce Bridge model plugins. |
| Accelerator and rollout structure | Shared accelerator APIs and the rollout `engine_group.py` split centralize device operations and engine management. | Register the NPU accelerator, order training bootstrap correctly, and carry NPU resource/visible-device differences into the new engine-group structure without duplicating rollout orchestration. |
| Training/serving memory lifecycle | Shared training contexts and serving sleep/wake coordinate memory reuse between training and generation. | Initialize NPU optimizer state inside the existing training memory context; fix packed IPC reference ownership and reclamation; keep serving block-table/control tensors outside releasable KV-cache storage. |
| MTP updates | The shared flow updates the main model and draft model through separate weight-update transactions. | Add the Ascend worker's draft-target lifecycle and GLM MTP graph-compatible position mask, keeping the common Vime update flow unchanged. |

Native loading here refers to the model construction/loading/export path.
The NPU image still contains Bridge-related dependencies; removing all of those
packages is not part of this PR.

## Additional Ascend integration

- Integrate standalone MegatronAdaptor and TransformerEngineNPU, with the common Megatron patch applied before the NPU-specific patch.
- Rebase NPU vendor patches onto the paired serving sources: vLLM `e6bfe03a` and vLLM-Ascend `fd815467`.
- Make torch_dist conversion/reference loading the default Qwen3-30B-A3B test path; retain HF-only testing via `VIME_TEST_TORCH_DIST_REF_LOAD=0`.
- Add GLM-4.7-Flash non-colocate coverage. MTP is off by default and explicitly enabled 
- Update existing mapping, platform and launch-contract tests, remove obsolete CLI arguments, and keep local model/dataset/checkpoint overrides out of the commits.

## Validation

Local E2E results collected during integration:

| Model / mode | Coverage | Result |
| --- | --- | --- |
| Qwen3-4B | Non-colocate stateful HCCL weight updates, training and rollout | PASS |
| Qwen3-30B-A3B, HF | Colocate packed NPU IPC, rank-local MoE, training and rollout | PASS |
| Qwen3-30B-A3B, torch_dist | Checkpoint conversion, reference loading, training and rollout | PASS |
| Qwen3-VL-8B | Native loading, training TP4/SP, rollout TP1 per slot | PASS |
| GLM-4.7-Flash, no MTP | Native loading and non-colocate HCCL | PASS  |
| GLM-4.7-Flash, MTP eager | Two training steps, main/draft updates and subsequent generation | PASS |
| GLM-4.7-Flash, MTP graph | Actual graph capture, two training steps and main/draft updates | PASS |

## Not yet adapted / follow-ups

- **Qwen3.5 NPU:** the tested environment has incompatible serving/training GDN operator stacks: overlapping FLA/serving operator libraries and a serving causal-convolution Python-to-binary schema mismatch. It needs a matched serving binary/OPP and training FLA environment before completing NPU adaptation and E2E validation. Main's model implementation remains available.
- GLM colocate, other unvalidated model/MTP combinations, Omni, DSpark, speculative metrics and pull/delta weight transfer are outside the validated NPU scope.
- Broader parallel configurations, long-running memory stability and performance equivalence still need dedicated validation.
```

### CODE REVIEW

File: docker/amd_patch/latest/amd_megatron_fused_kernels_init.patch [hunk 1]
```diff
@@ -0,0 +1,51 @@
+diff --git a/megatron/legacy/fused_kernels/__init__.py b/megatron/legacy/fused_kernels/__init__.py
+index 87cceac3..ac686d74 100644
+--- a/megatron/legacy/fused_kernels/__init__.py
++++ b/megatron/legacy/fused_kernels/__init__.py
```
Comment (momo609) [inline docker/amd_patch/latest/amd_megatron_fused_kernels_init.patch:4]: amd patch can remove

File: docker/npu_patch/megatron.patch [hunk 2]
```diff
@@ -455,4906 +455,29 @@ index dd8590a79..3736b0e31 100644
              assert len(steps) == 1
              step = torch.tensor(steps[0], dtype=torch.float)
  
-diff --git a/megatron/core/ssm/chunk_gated_delta_rule.py b/megatron/core/ssm/chunk_gated_delta_rule.py
-new file mode 100644
-index 000000000..dc388fd7f
---- /dev/null
-+++ b/megatron/core/ssm/chunk_gated_delta_rule.py
```
Comment (momo609) [inline docker/npu_patch/megatron.patch:462]: The GDN Triton operator has been removed. Is that because Qwen3.5’s training side is already supported in another third-party repo?

**Comment by Meihan-chen** [comment]
qwen3-4B non-colocate
image



qwen3-30B-A3B colocate  vs  https://github.com/vllm-project/vime/pull/269
image

File: docker/amd_patch/latest/amd_megatron_fused_kernels_init.patch (hunk 1 above)
Comment (Meihan-chen) [reply docker/amd_patch/latest/amd_megatron_fused_kernels_init.patch]: done

**Review by momo609 — APPROVED** [review]

File: .buildkite/npu_suites.py [hunk 3]
```diff
@@ -31,7 +31,6 @@
         ("test_qwen3_4B_npu.py", "npu-8", "", {}),
         ("test_qwen3_30B_A3B_npu.py", "npu-16", "", {}),
         ("test_qwen3_vl_8B_npu.py", "npu-8", "", {}),
-        ("test_qwen3.5_35B_A3B_npu.py", "npu-16", "", {}),
```
Comment (CalvinXKY) [inline .buildkite/npu_suites.py:34]: Why do we remove qwen3.5?

Comment (Meihan-chen) [reply .buildkite/npu_suites.py:34]: Thanks for raising this, and thanks to Fulin for the original Qwen3.5 NPU support in #409. The other validated models have completed the migration, but Qwen3.5 needs additional work beyond moving from Bridge to main’s native HF loading/export flow. Its hybrid Attention/Gated DeltaNet architecture also requires reconnecting the specialized NPU GDN and convolution operators to main’s model implementation.

During the migration, we also found compatibility issues between the training FLA and serving operators, including conflicting operator registrations that need to be addressed in FLA-NPU. I haven’t been able to get the full-model flow passing yet, so GDN forward/backward execution, and post-training weight synchronization still need validation within the complete flow. 

I therefore temporarily removed the corresponding NPU test and CI entry from this PR.  I’d like to address this in a follow-up and would really appreciate @Fulin-Gao ’s help with the remaining Qwen3.5 NPU adaptation, given his familiarity with the original implementation. Thanks!
